cli/download_source.go

```diff
@@ -325,31 +325,68 @@ func isLocalSource(sourceUrl *url.URL) bool {
-// (the terraform init command will redownload the latest ones), but leave all the other files, such
-// as the .terraform folder with the downloaded modules and remote state settings.
-func cleanupTerraformFiles(path string, terragruntOptions *options.TerragruntOptions) error {
-	if !util.FileExists(path) {
+// If this download dir already exists, delete its contents to ensure that we don't have any old files from previous
+// runs. Leave only the non-Terraform files in the working dir, as that dir may contain terraform state we don't want
+// to lose / reconfigure as well as module and provider code we don't want to re-download.
+func cleanupDownloadDir(terraformSource *TerraformSource, terragruntOptions *options.TerragruntOptions) error {
+	if !util.FileExists(terraformSource.DownloadDir) {
 		return nil
 	}
 
-	terragruntOptions.Logger.Printf("Cleaning up existing *.tf files in %s", path)
+	terragruntOptions.Logger.Printf("Cleaning up contents of download folder %s", terraformSource.DownloadDir)
 
-	files, err := zglob.Glob(util.JoinPath(path, "**/*.tf"))
+	// Backup the working dir if it exists
+	backupWorkingDirFolder := util.JoinPath(terragruntOptions.DownloadDir, ".working-dir-backup")
+	if util.FileExists(terraformSource.WorkingDir) {
+		if err := os.MkdirAll(terragruntOptions.DownloadDir, 0700); err != nil {
+			return errors.WithStackTrace(err)
+		}
+		if err := os.Rename(terraformSource.WorkingDir, backupWorkingDirFolder); err != nil {
+			return errors.WithStackTrace(err)
+		}
+	}
+
+	// Delete everything that was in the download dir
+	files, err := zglob.Glob(util.JoinPath(terraformSource.DownloadDir, "*"))
 	if err != nil {
 		return errors.WithStackTrace(err)
 	}
+	if err := util.DeleteFilesAndFolders(files); err != nil {
+		return err
+	}
 
-	// Filter out files in .terraform folders, since those are from modules downloaded via a call to terraform get,
-	// and we don't want to re-download them.
-	filteredFiles := []string{}
-	for _, file := range files {
-		if !strings.Contains(file, ".terraform") {
-			filteredFiles = append(filteredFiles, file)
+	// Restore the working dir and delete any Terraform code in it
+	if util.FileExists(backupWorkingDirFolder) {
+		if err := os.MkdirAll(filepath.Dir(terraformSource.WorkingDir), 0700); err != nil {
+			return errors.WithStackTrace(err)
+		}
+
+		if err := os.Rename(backupWorkingDirFolder, terraformSource.WorkingDir); err != nil {
+			return errors.WithStackTrace(err)
+		}
+
+		// Find any Terraform files in the working dir
+		terraformFiles, err := zglob.Glob(util.JoinPath(terraformSource.WorkingDir, "**/*.tf"))
+		if err != nil {
+			return errors.WithStackTrace(err)
+		}
+		filteredTerraformFiles := []string{}
+
+		// Filter out files in .terraform folders, since those are from modules downloaded via a call to terraform init,
+		// and we don't want to re-download them.
+		for _, terraformFile := range terraformFiles {
+			if !strings.Contains(terraformFile, ".terraform") {
```

Comment:
Should this actually look at the path components? Is there a situation where you could have a hidden file that we want to actually delete, like `.terraform-test.tf`?

Reply:
Yes, I noticed that and felt too lazy to do it right... But, fine, _grumble grumble_. https://github.com/gruntwork-io/terragrunt/pull/739/commits/7c8a472ac48bfa9907b4a33502f74062ffff677d